doc/TWAMP/SAI-Proposal-TWAMP-LIGHT.md

```diff
@@ -0,0 +1,1085 @@
+
+## TWAMP Light offloading in ASIC ##
+Since TWAMP Light is a simple and lightweight protocol, it is easy for vendors to implement TWAMP Light in their ASIC, which can improve performance and reduce CPU overload on network devices. TWAMP Light offloading in ASIC should work for sending and receiving TWAMP Light packets, collecting measurement data such as latency, jitter and packet loss. And also, RFC-5357 TWAMP Light should be followed. For TWAMP Light hardware solution, the main requirements are as follows:
+* Can be deployed as one sender
+  * Construct Test-request packets based on configuration including IP header, UDP port, etc.
+  * Transmit Test-request packets to a reflector within tx timestamps which are captured in ASIC.
+  * Receive Test-response packets from a reflector and capture rx timestamps in ASIC.
+* Can be deployed as one reflector
+  * Receive Test-request packets from a sender and capture rx timestamps in ASIC.
+  * Construct Test-response packets based on configuration including IP header, UDP port, etc.
+  * Encode rx timestamps of Test-request packets into Test-response packets.
+  * Reflect Test-response packets back to a sender within tx timestamps which are captured in ASIC.
+* Can collect measurement results
+  * Collect measurement data and calculate point-to-point latency/jitter/packet loss rate
+  * Report the latency/jitter/packet loss rate to software
+
+## TWAMP Light SAI components ##
+TWAMP Light can be broken down into following components for configuring twamp Light in ASIC.
+* Global Information
+  * Resource of TWAMP Light in ASIC
+  * Notification function of state and statistics
+* TWAMP Light session
+  * Session mode
+  * Session role
+  * Session state
+  * Authentication mode
+  * Virtual Router
+  * Packet encapsulation type
+  * Packet forwarding mode
+* TWAMP Light test packets
+  * IP source address
+  * IP destination address
+  * IP TTL
+  * IP DSCP
+  * UDP source port
+  * UDP destination port
+  * Format of timestamp
+* TWAMP Light measurement controlling
+  * Packets measurement mode
+  * Starting and stopping to transmit packets
+  * Interval of transmitting packets
+  * Interval of getting measurement data
+  * Timeout of receiving packets
+* TWAMP Light measurement data
+  * Rx packets count
+  * Rx packets bytes
+  * Tx packets count
+  * Tx packets bytes
+  * Drop packets count
+  * Maximum latency
+  * Minimum latency
+  * Average latency
+  * Maximum jitter
+  * Minimum jitter
+  * Average jitter
+
+## Measurement mode ##
+
+Following diagram introduces the measurement mode of TWAMP Light Session-Sender.
+
+![Packet-Count](./figures/TWAMP_Light_packet_count.png "Figure 2: Packet Count Mode")
+
+![Period](./figures/TWAMP_Light_period.png "Figure 2: Period Mode")
+
+![Continuous](./figures/TWAMP_Light_continuous.png "Figure 2: Continuous Mode")
+
+## How to get measurement data ##
+It provides two methods to get measurement data for application.
```

Comment:
what is the way to configure one of them or both ?

Reply:
There are two methods to get measurement data. 
One is that the sai plugin proactively reports measurement data periodically through the callback function called "sai_twamp_session_event_notification_data_t". It is the default method and should be implemented and supported by sai plugin.
The other is that the application can actively retrieve measurement data through "sai_get_twamp_session_stats_fn" or "sai_get_twamp_session_stats_ext_fn" when the application wants to get the  measurement data immediately. It is an additional and optional method.
So these two methods are not conflict when they are both supported. The first one is defaultly supported by sai plugin. The other is optional and depends on the appliction.

Reply:
The SAI Callback function "sai_twamp_session_state_change_notification_fn" in MD file has been modified as "sai_twamp_session_event_notification_data_t" based on the file "saitwamp.h"